rs/bitcoin/kyt/src/lib.rs

```diff
@@ -92,33 +75,114 @@ async fn get_tx(tx_id: String) -> Result<Transaction, BitcoinTxError> {
-            let decoded_tx_id = tx.compute_txid().to_string();
-            if decoded_tx_id != tx_id {
-                return Err(BitcoinTxError::TxIdMismatch {
-                    expected: tx_id,
-                    decoded: decoded_tx_id,
+            let decoded_tx_id = tx.compute_txid();
+            if decoded_tx_id.as_ref() as &[u8; 32] != tx_id.as_ref() {
+                return Err(GetTxError::TxIdMismatch {
+                    expected: tx_id.into(),
+                    decoded: *decoded_tx_id.as_ref(),
                 });
             }
             Ok(tx)
         }
+        Err((r, m)) if is_response_too_large(&r, &m) => Err(GetTxError::ResponseTooLarge),
         Err((r, m)) => {
             // TODO(XC-158): maybe try other providers and also log the error.
             println!("The http_request resulted into error. RejectionCode: {r:?}, Error: {m}");
-            Err(BitcoinTxError::Rejected {
+            Err(GetTxError::Rejected {
                 code: r,
                 message: m,
             })
         }
     }
 }
 
+impl HasOutPoint for TxIn {
+    fn txid(&self) -> Txid {
+        Txid::from(*(self.previous_output.txid.as_ref() as &[u8; 32]))
+    }
+    fn vout(&self) -> u32 {
+        self.previous_output.vout
+    }
+}
+
+impl TransactionLike for Transaction {
+    type Input = TxIn;
+    type AddressError = GetTxError;
+
+    fn iter_inputs(&self) -> impl Iterator<Item = &TxIn> {
+        self.input.iter()
+    }
+    fn output_address(&self, vout: u32) -> Result<Address, GetTxError> {
+        let output = &self.output[vout as usize];
+        Address::from_script(&output.script_pubkey, Network::Bitcoin).map_err(GetTxError::Address)
+    }
+}
+
+struct KytCanisterState;
+
+impl FetchState<Transaction> for KytCanisterState {
+    type FetchGuard = state::FetchGuard;
+
+    fn new_fetch_guard(&self, txid: Txid) -> Result<Self::FetchGuard, FetchGuardError> {
+        state::FetchGuard::new(txid)
+    }
+
+    fn get_fetch_status(&self, txid: Txid) -> Option<FetchTxStatus<Transaction>> {
+        state::get_fetch_status(txid)
+    }
+
+    fn set_fetch_status(&self, txid: Txid, status: FetchTxStatus<Transaction>) {
+        state::set_fetch_status(txid, status)
+    }
+
+    fn set_fetched_address(&self, txid: Txid, index: usize, address: Address) {
+        state::set_fetched_address(txid, index, address)
+    }
+}
+
+struct KytCanisterEnv;
+
+impl FetchEnv<TxIn, Transaction> for KytCanisterEnv {
+    async fn get_tx(&self, txid: Txid, buffer_size: u32) -> Result<Transaction, GetTxError> {
+        get_tx(txid, buffer_size).await
+    }
+    fn cycles_accept(&self, cycles: u128) -> u128 {
+        ic_cdk::api::call::msg_cycles_accept128(cycles)
+    }
+    fn cycles_available(&self) -> u128 {
+        ic_cdk::api::call::msg_cycles_available128()
+    }
+}
+
+pub async fn check_transaction_inputs(txid: Txid) -> CheckTransactionResponse {
```

Comment:
@ninegua @THLO I have a general understanding question, maybe it's better do quickly discuss it offline. Why doesn't the canister use timers?

My understanding is that now if there is an HTTPs outcall that result in an error (e.g. consensus error), this error would be returned to the caller and it's up to the caller to call the service again (and pay the fee again). As a user I think it would be more understandable if the canister would retry in case of an internal failure. WDYT?